File: benchmark/comparison/results/haiku_py/08.py
class Expr:
    pass

class Num(Expr):
    def __init__(self, n):
        self.n = n

class Add(Expr):
    def __init__(self, a, b):
        self.a = a
        self.b = b

class Mul(Expr):
    def __init__(self, a, b):
        self.a = a
        self.b = b

class Neg(Expr):
    def __init__(self, x):
        self.x = x

def eval_expr(expr):
    if isinstance(expr, Num):
        return expr.n
    elif isinstance(expr, Add):
        return eval_expr(expr.a) + eval_expr(expr.b)
    elif isinstance(expr, Mul):
        return eval_expr(expr.a) * eval_expr(expr.b)
    elif isinstance(expr, Neg):
        return -eval_expr(expr.x)

def show(expr):
    if isinstance(expr, Num):
        return str(expr.n)
    elif isinstance(expr, Add):
        return f"({show(expr.a)} + {show(expr.b)})"
    elif isinstance(expr, Mul):
        return f"({show(expr.a)} * {show(expr.b)})"
    elif isinstance(expr, Neg):
        return f"(-({show(expr.x)}))"

# Test eval
print(eval_expr(Num(5)))
print(eval_expr(Add(Num(3), Num(5))))
print(eval_expr(Mul(Num(2), Num(4))))
print(eval_expr(Neg(Num(3))))
print(eval_expr(Neg(Add(Num(2), Num(3)))))

# Test show
print(show(Num(5)))
print(show(Add(Num(3), Num(5))))
print(show(Mul(Num(2), Num(4))))
print(show(Neg(Num(3))))
print(show(Neg(Add(Num(2), Num(3)))))

File: benchmark/comparison/results/adv_ts/08.ts
type Expr =
  | { tag: "Num"; value: number }
  | { tag: "Add"; a: Expr; b: Expr }
  | { tag: "Mul"; a: Expr; b: Expr }
  | { tag: "Neg"; x: Expr };

function num(n: number): Expr { return { tag: "Num", value: n }; }
function add(a: Expr, b: Expr): Expr { return { tag: "Add", a, b }; }
function mul(a: Expr, b: Expr): Expr { return { tag: "Mul", a, b }; }
function neg(x: Expr): Expr { return { tag: "Neg", x }; }

function eval_expr(e: Expr): number {
  switch (e.tag) {
    case "Num": return e.value;
    case "Add": return eval_expr(e.a) + eval_expr(e.b);
    case "Mul": return eval_expr(e.a) * eval_expr(e.b);
    case "Neg": return -eval_expr(e.x);
  }
}

function show(e: Expr): string {
  switch (e.tag) {
    case "Num": return String(e.value);
    case "Add": return `(${show(e.a)} + ${show(e.b)})`;
    case "Mul": return `(${show(e.a)} * ${show(e.b)})`;
    case "Neg": return `(-(${show(e.x)}))`;
  }
}

// eval tests
console.log(eval_expr(num(5)));
console.log(eval_expr(add(num(3), num(5))));
console.log(eval_expr(neg(num(3))));
console.log(eval_expr(neg(add(num(2), num(3)))));

// show tests
console.log(show(num(5)));
console.log(show(add(num(3), num(5))));
console.log(show(mul(num(2), num(4))));
console.log(show(neg(num(3))));
console.log(show(neg(add(num(2), num(3)))));
